share/completions/git.fish

```diff
@@ -1,5 +1,32 @@
 # fish completion for git
-# Use 'command git' to avoid interactions for aliases from git to (e.g.) hub
+
+# Use this instead of calling git directly; it passes the commands that are
+# already present on the commandline to git. This is relevant for --work-tree etc, see issue #6219.
+function __fish_git
+    set -l saved_args $argv
+    set -l global_args
+    set -l cmd (commandline -opc)
+    # We assume that git is the first command until we have a better awareness of subcommands, see #2705.
+    set -e cmd[1]
+    if argparse -s (__fish_git_global_optspecs) -- $cmd 2>/dev/null
+        set -l subcommand $argv[1]
+        if set -q subcommand
+            for arg in $cmd
+                test $arg = $subcommand; and break
+                set -a global_args $arg
+            end
+        end
+    end
+    # Using 'command git' to avoid interactions for aliases from git to (e.g.) hub
+    command git $global_args $saved_args
+end
+
+# Print an optspec for argparse to handle git's options that are independent of any subcommand.
+function __fish_git_global_optspecs
```

Comment:
Wait could this be a variable as well, are there any advantages of a function here?

Reply:
It could be a variable, the tricky thing there is scoping, because you want to share it with multiple functions.

So using a function is probably easier, if slightly slower (tho I doubt it matters).

Reply:
I assume that no one else uses a function/variable with the same name, otherwise it can break either way. But yeah, it shouldn't matter.